controllers/badge.js

```diff
@@ -0,0 +1,661 @@
+  constructor() {
+    super(true);
+  }
+
+  /**
+   * function to get userId from firebase token
+   *
+   * @param {*}
+   *        event: event triggers that cause the invocation of the lambda
+   *
+   * @returns the userId
+   */
+
+  getUserId = async (event) => {
+    let userId;
+    const FIREBASE_ENDPOINT = await getSecretFun.getSecrets('FirebaseEndpoint');
+    const idToken = event.headers.Authorization.split(BLANK_SPACE)[1];
+    return await axios
+      .post(FIREBASE_ENDPOINT, {
+        firebase_token: idToken,
+      })
+      .then((response) => {
+        userId = response.data.data.cst_uid;
+        return userId;
+      });
+  };
+
+  /**
+   * function to get user's login and signup data from firebase token
+   *
+   * @param {*}
+   *        event: event triggers that cause the invocation of the lambda
+   *
+   * @returns an object of login and signup info
+   */
+  getUserData = async (event) => {
+    let result;
+    const idToken = await event.headers.Authorization;
+    const url = await getSecretFun.getSecrets('UserdataEndpoint');
+    return await axios.get(url, {
+      headers: { Authorization: idToken },
+      params: { info_params: `['login_info']` }
+    }).then((response) => {
+      result = response.data.response;
+      return result;
+    })
+  }
+
+  /**
+   * function to give a badge to the user
+   * @param {*}
+   *      event: event triggers that cause the invocation of the lambda
+   *      conn: MongoDB connection object
+   *
+   */
+  assignBadge = async (event, conn) => {
+    this.connection = this.connection || conn;
+    let userId = await this.getUserId(event);
+    const userData = await this.getUserData(event);
+    // let userId = "10456";
+    // const userData = {
+    //   login_info: { user_id: '10456', sign_up: 1418157734, login_date: 1670044704 }
+    // };
+    const data = await this.connection
+      .collection(COLLECTIONS.STREAKS)
+      .find({
+        ...{ published_at: { $ne: null } },
+      })
+      .toArray();
+
+    if (data && data.length) {
+      for (let i = 0; i < data.length; i++) {
+        const videoIds = data[i].free_lessons,
+          quizIds = data[i].free_quizs,
+          chapterIds = data[i].chapters;
+
+        let isVideosCompleted = true, isQuizzesCompleted = true, isChaptersCompleted = true;
+        if (videoIds && videoIds.length) {
+          for (let j = 0; j < videoIds.length; j++) {
+            const progressData = await this.connection
+              .collection(COLLECTIONS.USER)
+              .find({
+                componentId: videoIds[j],
+                userId,
+                date: {
+                  $lte: userData.login_info.sign_up + (data[i].streakNo * 604800000)
+                },
+                'progressData.status': 'completed',
+              })
+              .toArray();
+
+            if (!progressData || !progressData.length) {
+              isVideosCompleted = false;
+            }

```

Comment:
if progressData of videos, has even 1 completed video 
isVideosCompleted will remain  = true